src/phasorpy/phasor.py

```diff
@@ -34,23 +35,216 @@
+    dtype : dtype_like, optional
+        Data type of output arrays. Either float32 or float64 (default).
+    num_threads : int, optional
+        Number of OpenMP threads to use for parallelization.
+        By default, multi-threading is disabled.
+        If zero, up to half of logical CPUs are used.
+        OpenMP may not be available on all platforms.
+
+    Returns
+    -------
+    mean : ndarray
+        Average of signal along axis.
+    real : ndarray
+        Real component of phasor coordinates along axis.
+    imag : ndarray
+        Imaginary component of phasor coordinates along axis.
+
+    Raises
+    ------
+    ValueError
+        The `signal` has less than three samples along `axis` or the
+        `sample_phase` size does not equal the number of samples along `axis`.
+    TypeError
+        The `signal` or `dtype` types are not supported.
+
+    Examples
+    --------
+    Calculate phasor coordinates of a phase-shifted sinusoidal waveform:
+
+    >>> sample_phase = numpy.linspace(0, 2 * math.pi, 5, endpoint=False)[::-1]
+    >>> signal = 1.1 * (
+    ...     numpy.cos(sample_phase - 0.78539816) * 2 *  0.70710678 + 1
+    ... )
+    >>> phasor_from_signal_f1(
+    ...    signal, sample_phase=sample_phase
+    ... )  # doctest: +NUMBER
+    (1.1, 0.5, 0.5)
+
+    """
+    signal = numpy.array(signal, order='C', ndmin=1, copy=False)
+    # assert signal.flags['C_CONTIGUOUS']
+    samples = signal.shape[axis]  # this also verifies axis
+    if samples < 3:
+        raise ValueError(f'not enough {samples=} along {axis=}')
+    if sample_phase is None:
+        sample_phase = numpy.linspace(
+            0, 2 * math.pi, samples, endpoint=False, dtype=numpy.float64
+        )
+    else:
+        sample_phase = numpy.array(
+            sample_phase, dtype=numpy.float64, copy=False, ndmin=1
+        )
+        if sample_phase.ndim != 1 or sample_phase.size != samples:
+            raise ValueError(f'{sample_phase.shape=} != ({samples},)')
+    if num_threads is None:
+        num_threads = 1
+    elif num_threads == 0:
+        try:
+            num_threads = len(os.sched_getaffinity(0))  # type: ignore
+        except AttributeError:
+            # sched_getaffinity not available on Windows
+            num_threads = os.cpu_count()
+            if num_threads is None:
+                num_threads = 1
+        num_threads = max(num_threads // 2, 1)
+    elif num_threads < 0:
+        raise ValueError(f'{num_threads=} < 0')
+
+    # pure numpy implementation for reference:
+    # shape = [1] * signal.ndim
+    # shape[axis] = sample_phase.size
+    # sample_phase = sample_phase.reshape(shape)  # make broadcastable
+    # mean = numpy.mean(signal, axis=axis)
+    # real = numpy.mean(signal * numpy.cos(sample_phase), axis=axis)
+    # real /= mean
+    # imag = numpy.mean(signal * numpy.sin(sample_phase), axis=axis)
+    # imag /= mean
+    # return mean, real, imag
```

Comment:
Should we keep this commented implementation?

Reply:
I was thinking to move it into `test_phasor.py` or `phasorpy.testing_utils.py`, but then the test function would need to be tested too. Also, the plan is to test `phasor_from_signal_f1` against `phasor_from_signal` once implemented, or to merge those functions into one...

Reply:
You are right, it can be a good a idea to keep that implementation somewhere to compare. The proposed `phasor_from_signal` is also different from that implementation, so maybe we can keep and compare the three of them?

Reply:
For now there is no need for this in the tests, which use known values. So I think to leave it as a comment in this PR. Another option would be to implement it as a reference in a tutorial/benchmark that compares various implementations.